### ISSUE TITLE
Crash when running with parallel HDF5 and Calculate_radial_profiles=1

### ISSUE THREAD

**Issue by jchelly** [ICRAR/VELOCIraptor-STF#24]
If Velociraptor is built with parallel HDF5 support and we have MPI_number_of_tasks_per_write>1 and Calculate_radial_profiles=1 in the .cfg file then the code crashes with the following message:
```
HDF5-DIAG: Error detected in HDF5 (1.10.3) MPI-process 15:
  #000: H5Dio.c line 322 in H5Dwrite(): could not get a validated dataspace from file_space_id
    major: Invalid arguments to routine
    minor: Bad value
  #001: H5S.c line 254 in H5S_get_validated_dataspace(): selection + offset not within extent
    major: Dataspace
    minor: Out of range
Failed to write dataset: Npart_profile
```

The problem seems to be in write_dataset_nd() in hdfitems.h. When parallel HDF5 is enabled one MPI communicator is created for each output file. All of the tasks in one communicator write to the same output file. Each task needs to calculate the offset at which it should write its data and this offset calculation is wrong - it results in tasks trying to write beyond the bounds of the dataset.

The offset is stored in dims_offset and calculated as follows (this occurs several times in hdfitems.h):
```
            MPI_Allgather(dims_single.data(), rank, MPI_UNSIGNED_LONG_LONG, mpi_hdf_dims.data(), rank, MPI_UNSIGNED_LONG_LONG, comm);
            MPI_Allreduce(dims_single.data(), mpi_hdf_dims_tot.data(), rank, MPI_UNSIGNED_LONG_LONG, MPI_SUM, comm);
            for (auto i=0;i<rank;i++) {
                dims_offset[i] = 0;
                if (flag_first_dim_parallel && i > 0) continue;
                for (auto j=1;j<=ThisWriteTask;j++) {
                    dims_offset[i] += mpi_hdf_dims[i*NProcs+j-1];
                }
            }
```
The dimensions of the arrays on all tasks are gathered in mpi_hdf_dims and used to compute the offset in each dimension for the data to be written by this task (dims_offset). Here I think the index into mpi_hdf_dims is wrong. I think
```
                for (auto j=1;j<=ThisWriteTask;j++) {
                    dims_offset[i] += mpi_hdf_dims[i*NProcs+j-1];
                }
```
should really be
```
                for (auto j=0;j<ThisWriteTask;j++) {
                    dims_offset[i] += mpi_hdf_dims[j*rank+i];
                }
```
because after the allgather mpi_hdf_dims contains rank elements from each MPI task. j is looping over the lower numbered MPI tasks and i is the index within the block received from each task.

This would mean that any multidimensional output arrays will either be corrupted or cause a crash.

**Comment by rtobar** [comment]
On top of the, there are two versions of `write_dataset_nd`, one taking `void *data` and another taking `T *data`, with `T` being a template parameter. We should probably start by removing the duplicate and implementing one in terms of the other.

**Comment by rtobar** [comment]
@jchelly you read my mind :+1:

**Comment by jchelly** [comment]
I've been looking into this a bit more to understand why it hasn't caused problems before now, given that the same function is used to write 1D datasets in parallel mode and the example config files enable output of density profiles (which are 2D datasets).

In the 1D case rank=1 and i=0 always and the old code is equivalent to the fixed version. When it's writing 2D datasets output gets written at the wrong location in the file but I think you have to be quite unlucky to get the out of bounds error - it's more likely that the offsets are too small so density profiles (for example) get written in the wrong location and/or overwritten and the last part of the dataset is left zeroed out but the code doesn't crash.

**Comment by jchelly** [comment]
This is fixed by #26, which is now in master.
